### PR TITLE
server: support decoding prepared string args to character_set_client

### LINKED ISSUE #30722: string args of `COM_STMT_EXECUTE` are not aware of character_set_client
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```go
package main

import (
	"context"
	"database/sql"
	"fmt"

	_ "github.com/go-sql-driver/mysql"
)

func main() {
	ctx := context.Background()
	db, err := sql.Open("mysql", "root:@tcp(127.0.0.1:4000)/test")
	if err != nil {
		panic(err.Error())
	}
	conn, err := db.Conn(ctx)
	if err != nil {
		panic(err.Error())
	}
	_, _ = conn.ExecContext(ctx, "set names gbk;")
	_, _ = conn.ExecContext(ctx, "drop table if exists t;")
	_, _ = conn.ExecContext(ctx, "create table t (a char(255));")
	stmt, err := conn.PrepareContext(ctx, "insert into t values (?);")
	if err != nil {
		panic(err.Error())
	}
	_, err = stmt.Exec("'����'")
	if err != nil {
		panic(err.Error())
	}
	_, _ = conn.ExecContext(ctx, "set names utf8mb4;")
	rs, err := conn.QueryContext(ctx, "select * from t;")
	if err != nil {
		panic(err.Error())
	}
	defer rs.Close()
	for rs.Next() {
		var s string
		if err := rs.Scan(&s); err != nil {
			panic(err.Error())
		}
		fmt.Println(s)
	}
}

```

### 2. What did you expect to see? (Required)
```
'锟斤拷锟斤拷'
```

### 3. What did you see instead (Required)

```
'����'
```

### 4. What is your TiDB version? (Required)

6c0fcea07

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Fix #30722.

Problem Summary:

Previously, the command `ComStmtExecute` is not aware of `@@character_set_client`.

### What is changed and how it works?

This PR support 

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: server/conn_stmt.go [hunk 1]
```diff
@@ -310,14 +313,18 @@ func parseStmtFetchCmd(data []byte) (uint32, uint32, error) {
...
+func parseExecArgs(sc *stmtctx.StatementContext, args []types.Datum, boundParams [][]byte,
+	nullBitmap, paramTypes, paramValues []byte, enc *inputDecoder) (err error) {
 	pos := 0
 	var (
 		tmp    interface{}
 		v      []byte
 		n      int
 		isNull bool
 	)
+	if enc == nil {
```
Comment (zimulala) [inline server/conn_stmt.go:325]: When we will go this logic？It seems that if `chs` is "", enc is `UTF8Encoding`.

**Comment by zimulala** [comment]
PTAL  @xiongjiwei @Defined2014

File: server/conn_stmt.go (hunk 1 above)
Comment (tangenta) [reply server/conn_stmt.go:325]: This is used to prevent the misuse of this function.

**Review by zimulala — APPROVED** [review]
LGTM

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**